File: neet_code_150_cpp/bit_manipulation/reverse_bits.cpp
#include <cstdint>
#include <iostream>
#include <stack>
#include <system_error>

class Solution {
    public:
        // AC
        uint32_t reverseBits(uint32_t n) {
            uint32_t result = 0;
            std::stack<int> bit_list;

            for(int i = 0; i < 32; ++i) {
                if(n & (1 << i)) {
                    bit_list.push(1);
                } else {
                    bit_list.push(0);
                }
            }

            int i = 0;
            while(!bit_list.empty()) {
                result += bit_list.top() * (1 << i);
                bit_list.pop();
                ++i;
            }

            return result;
        }

        // AC
        uint32_t reverseBits2(uint32_t n) {
            uint32_t result = 0;

            int b = 0;
            for(int a = 31; a >= 0; --a) {
                if(n & (1 << a)) {
                    result += 1 << b;
                }
                ++b;
            }

            return result;
        }
};

// 模範解答
class SolutionAns {
public:
    uint32_t reverseBits(uint32_t n) {
        uint32_t result = 0;
        
        for (int i = 0; i < 32; i++) {
            result <<= 1;
            result |= n & 1;
            n >>= 1;
        }
        
        return result;
    }
};

int main(void) {
    uint32_t case_1 = 0b00000010100101000001111010011100;
    // => 0b00111001011110000010100101000000 (964176192)
    uint32_t case_2 = 0b11111111111111111111111111111101;
    // => 0b10111111111111111111111111111111 (3221225471)

    Solution s_1;
    std::cout << s_1.reverseBits(case_1) << std::endl;
    std::cout << s_1.reverseBits(case_2) << std::endl;

    std::cout << s_1.reverseBits2(case_1) << std::endl;
    std::cout << s_1.reverseBits2(case_2) << std::endl;

    SolutionAns s_ans;
    std::cout << s_ans.reverseBits(case_1) << std::endl;
    std::cout << s_ans.reverseBits(case_2) << std::endl;
}

File: neet_code_150_go/bit_manipulation/reverse_bits.go
package main

import "fmt"

// AC
func reverseBits(num uint32) uint32 {
	bit_list := [32]int{}
	for i := 0; i < 32; i++ {
		if num&(1<<i) != 0 {
			bit_list[i] = 1
		}
	}
	// fmt.Println("bit_list: ", bit_list)

	result := uint32(0)
	for i := 31; i >= 0; i-- {
		result += uint32(bit_list[i]) * (1 << (31 - i))
	}

	return result
}

// AC
func reverseBits2(num uint32) uint32 {
	result := uint32(0)
	for i := 31; i >= 0; i-- {
		if num&(1<<i) != 0 {
			result += 1 << (31 - i)
		}
	}

	return result
}

// AC
// 模範解答より
func reverseBits3(num uint32) uint32 {
	result := uint32(0)
	for i := 0; i < 32; i++ {
		result = (result << 1) | (num & 1)
		num >>= 1
	}

	return result
}

func main() {
	case_1 := uint32(0b00000010100101000001111010011100)
	// => 00111001011110000010100101000000
	case_2 := uint32(0b11111111111111111111111111111101)
	// => 10111111111111111111111111111111

	fmt.Printf("case_1: %b\n", reverseBits(case_1))
	fmt.Printf("case_2: %b\n", reverseBits(case_2))

	fmt.Printf("case_1: %b\n", reverseBits2(case_1))
	fmt.Printf("case_2: %b\n", reverseBits2(case_2))

	fmt.Printf("case_1: %b\n", reverseBits3(case_1))
	fmt.Printf("case_2: %b\n", reverseBits3(case_2))
}
